outlines/text/sequences/sequence.py

```diff
@@ -0,0 +1,238 @@
+        batch_shape_finished = token_ids_finished.shape[:-1]
+        token_ids_finished = np.concatenate(
+            [
+                token_ids_finished,
+                self.model.tokenizer.pad_token
+                * np.ones(batch_shape_finished + (1,), dtype=np.int64),
+            ],
+            axis=-1,
+        )
+
+        new_token_ids[~is_finished] = token_ids_unfinished
+        new_token_ids[is_finished] = token_ids_finished
+
+        return new_token_ids
+
+    def __call__(
+        self, prompt: Union[str, List[str]], samples: int = 1
+    ) -> Union[str, List[str]]:
+        """Generate a new sequence given a prompt.
+
+        Parameters
+        ----------
+        prompt
+            The input prompt.
+        samples
+            The number of samples to generate for each prompt.
+
+        Returns
+        -------
+        The full sequence that contains the prompts and the generated string.
+
+        """
+        token_ids, attention_mask = self.model.tokenizer.encode(prompt)
+        num_prompt_tokens = token_ids.shape[-1]
+
+        if samples > 1:
+            token_ids, _ = self.step(token_ids, attention_mask, samples)
+            is_finished = self.is_finished(token_ids)
+
+            num_batch_dims = token_ids.ndim - 1
+            repetitions = (samples,) + (1,) * num_batch_dims
+            attention_mask = np.tile(attention_mask, repetitions)
+            attention_mask = self.expand_attention_mask(attention_mask)
+        else:
+            batch_shape = token_ids.shape[:-1]
+            is_finished = np.zeros(batch_shape, dtype=np.bool_)
+
+        while True:
+            num_generated_tokens = token_ids.shape[-1] - num_prompt_tokens
+            if np.all(is_finished) or num_generated_tokens == self.max_tokens:
+                break
+
+            token_ids_unfinished = token_ids[~is_finished]
+            attention_mask_unfinished = attention_mask[~is_finished]
+            token_ids_unfinished, _ = self.step(
+                token_ids_unfinished, attention_mask_unfinished
+            )
+
+            token_ids = self.update_token_ids(
+                is_finished, token_ids, token_ids_unfinished
+            )
+            attention_mask = self.expand_attention_mask(attention_mask)
+            is_finished[~is_finished] = self.is_finished(token_ids_unfinished).flatten()
+
+        result = self.model.tokenizer.decode(token_ids)
+
+        if len(result) == 1:
+            return result[0]
+
+        return result
+
+
+def vectorized_random_choice(
+    items: NDArray[np.int64],
+    p: NDArray[np.float64],
+    samples: int = 1,
+):
+    """Vectorized implementation of `np.random.choice`.
+
+    `np.random.choice` does not support arrays of probability. This implements
+    the equivalent of this function where the `p` argument can be a matrix.
+
+    Note
+    ----
+    `searchsorted` might be more efficient here since the number of elements
+    can be quite large.
```

Comment:
That's [what we used in Aesara](https://github.com/aesara-devs/aesara/blob/2368ed3e7f14801785280038094ac99773e876a6/aesara/tensor/random/basic.py#LL1970C26-L1970C26).

Reply:
Made the change. Note that despite the smaller complexity `searchsorted` is slower for small arrays, but here the vocabulary is large and I found appreciable speedups for size ~ 10k+.